cdap-hbase-compat-0.98/src/main/java/co/cask/cdap/data2/transaction/messaging/coprocessor/hbase98/MessageTableRegionObserver.java

```diff
@@ -0,0 +1,162 @@
+import org.apache.hadoop.hbase.regionserver.Store;
+import org.apache.hadoop.hbase.regionserver.StoreScanner;
+import org.apache.hadoop.hbase.regionserver.compactions.CompactionRequest;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * RegionObserver for the Message Table of Transactional Messaging Service.
+ *
+ * This region observer does row eviction during flush time and compact time by using Time-To-Live (TTL) information
+ * in MetadataTable to determine if a message has expired its TTL.
+ *
+ * TODO: Add logic to remove data corresponding to invalidated transactions
+ */
+public class MessageTableRegionObserver extends BaseRegionObserver {
+
+  private static final Log LOG = LogFactory.getLog(MessageTableRegionObserver.class);
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_TYPE = new TypeToken<SortedMap<String, String>>() { }.getType();
+
+  private static final byte[] COL_FAMILY = HBaseTableFactory.COLUMN_FAMILY;
+  private static final byte[] COL = HBaseMetadataTable.COL;
+
+  private int prefixLength;
+  private LoadingCache<TopicId, TopicMetadata> topicCache;
+  private HTableInterface metadataTable;
+
+  @Override
+  public void start(CoprocessorEnvironment env) throws IOException {
+    if (env instanceof RegionCoprocessorEnvironment) {
+      HTableDescriptor tableDesc = ((RegionCoprocessorEnvironment) env).getRegion().getTableDesc();
+      prefixLength = Integer.parseInt(tableDesc.getValue(Constants.MessagingSystem.TABLE_PREFIX_BYTES));
+      String metadataTableNamespace = tableDesc.getValue(Constants.MessagingSystem.HBASE_METADATA_TABLE_NAMESPACE);
+      String metadataTableName = tableDesc.getValue(Constants.MessagingSystem.HBASE_METADATA_TABLE_NAME);
+      String hbaseNamespacePrefix = tableDesc.getValue(Constants.Dataset.TABLE_PREFIX);
+      HTableNameConverter nameConverter = new HTable98NameConverter();
+      metadataTable = env.getTable(nameConverter.toTableName(hbaseNamespacePrefix,
+                                                             TableId.from(metadataTableNamespace, metadataTableName)));
+      topicCache = CacheBuilder.newBuilder()
+        .expireAfterWrite(2, TimeUnit.MINUTES)
+        .maximumSize(1000)
+        .build(new CacheLoader<TopicId, TopicMetadata>() {
+
+          @Override
+          public TopicMetadata load(TopicId topicId) throws Exception {
+            Get get = new Get(MessagingUtils.toRowKeyPrefix(topicId));
+            Result result = metadataTable.get(get);
+            byte[] properties = result.getValue(COL_FAMILY, COL);
+            Map<String, String> propertyMap = GSON.fromJson(Bytes.toString(properties), MAP_TYPE);
+            return new TopicMetadata(topicId, propertyMap);
```

Comment:
We don't have to use `TopicMetadata` here. All we need is the "ttl" property. I am trying remove the dependency on cdap-tms.